docs/design/2021-08-18-charsets.md

```diff
@@ -0,0 +1,270 @@
+  - For internally executed SQL statements, since they are already utf-8, they do not need to be processed. For example, the table creation statement in the perfschema package.
+  
+### Runtime
+- Add a repertoire field to collationInfo to facilitate automatic character set conversion in expressions, so that many errors like "illegal mix of collations" can be avoided.
+  - The corresponding types of the Repertoire attribute are as follows:
+
+    ```go
+    type Repertoire int
+
+    const (
+      // RepertoireASCII is pure ASCII and it’s Unicode range: U+0000..U+007F
+      RepertoireASCII     Repertoire = 1           
+      // RepertoireExtended is Extended characters and it’s Unicode range: U+0080..U+FFFF    
+      RepertoireExtended  Repertoire = 1 << 1          
+      // RepertoireUnicode consists ASCII and EXTENDED, and it’s Unicode range: U+0000..U+FFFF 
+      RepertoireUnicode   Repertoire = ASCII | EXTENDED 
+    )
+    ```
+
+- Some of the built-in functions related to string computation require special processing after conversion from utf-8.
+  - Reference document: https://dev.mysql.com/doc/refman/8.0/en/string-functions.html.
+  - Handle the related functions in the Coprocessor.
+- Process some string-related internal functions, check whether the corresponding character set needs to be processed, specific functions such as DatumsToString, strToInt.
+
+### Optimizer
+
+- The statistics module may need special processing functions based on charset: AvgColSizeListInDisk, GetFixedLen, BucketToString and DecodedString, etc.
+- Ranger module
+  - The processing of prefix len needs to consider the charset.
+  - Functions such as BuildFromPatternLike and checkLikeFunc need to consider charset.
+
+### Collation
+
+Add gbk_chinese_ci and gbk_bin collations. In addition, considering the performance, we can add the collation of utf8mb4 (gbk_utf8mb4_bin).
+- Implement the Collator and WildcardPattern interface functions for each collation.
+  - gbk_chinese_ci and gbk_bin need to convert utf-8 to gbk encoding and then generate a sort key. gbk_utf8mb4_bin does not need to be converted to gbk code for processing.
+  - Implement the corresponding functions in the Coprocessor.
+
+### DDL
+
+Support for character sets of databases, tables and columns when creating databases, tables or adding columns, and support for changes through alter statements.
+
+### Others
+
+Other behaviors that need to be dealt with:
+- TiKVMinVersion information may need to be modified.
+- Set character set operations are supported, such as `set character set gbk`, `set names gbk` and `set character_set_client = gbk`, etc.
+- Support the correct display of character sets, such as show charset/collation.
+- load data: LoadDataInfo.getLine.
+
+### Compatibility
+
+#### Compatibility between TiDB versions
+
+- Upgrade compatibility:
+  - Upgrades from versions below 4.0 do not support gbk.
```

Comment:
```suggestion
  - Upgrades from versions below 4.0 do not support gbk or any character sets other than the original five (binary, ascii, latin1, utf8, utf8mb4).
```

what does "support" mean anyway? does this mean the following is rejected after the upgrade?

```sql
alter table below40 charset = gbk;
```

or does this mean the following is an error after the upgrade?

```sql
select _gbk'AAAA';
```

Reply:
Both statements should report errors.

Reply:
could you please include that in the document? ("any use of those character sets e.g. `select _gbk'AAAA';` will cause an error even after upgrading")

also that sounds very user-hostile. the only way a user can introduce charset to an existing cluster is by dumpling+lightning or binlog the entire database?